- GOAL      finish the pilot: bottom-up rotation, sign verdict, C3 v2, Calgary K=6, output stack, choropleth
- PRODUCED  bottom-up PCA on 21,111 DAs (cum var 86.9%), pop-weighted CMA scores, Stage 2 refit converged (20 coef, df.res −10), native-score map
- MISSED    map still flips (−0.946) on a seamless build: seam/reshape/pairing/predictors all cleared, reversal isolated to the outcome side; wrong-ruler twice, v1 truth_factors landmine; P4–P6 unrun
- NEXT      str the reduced_obj basis slot, read each city's curve at its own third knot vs mean truth; track → θ*-to-curve on trial; anti-track → Stage 1 inside

restore. lib cache → site-library, load pilot_session. list 07-22 tarball, expect 27 entries, ~19 fns, simulate_counts = exp version.

In [ ]:
DRIVE <- "/content/drive/MyDrive/thesis/dlnm-pilot"

system(sprintf("tar -xzf '%s/r_library.tar.gz' -C /content", DRIVE))
.libPaths(c("/content/site-library", .libPaths()))
suppressMessages({
  library(data.table); library(dlnm); library(gnm)
  library(splines);    library(mixmeta)
})

system(sprintf("tar -xzf '%s/saves_pilot_2026-06-03.tar.gz' -C /content", DRIVE))
pre <- ls()
load("/content/saves_pilot/pilot_session.RData")
loaded <- setdiff(ls(), c(pre, "pre"))

source(file.path(DRIVE, "fns.R"))

tar22 <- system(sprintf("tar -tzf '%s/saves_eod_2026-07-22.tar.gz'", DRIVE), intern = TRUE)

cat("tar 07-22 entries:", sum(grepl("\\.rds$", tar22)), "\n")
cat("session objects loaded:", length(loaded), "\n")
cat("key constants:", all(sapply(c("L","lag_weights","study_dates"), exists)), "\n")
cat("L dims:", paste(dim(L), collapse = " x "), "\n")
cat("fns present:", all(sapply(c("build_city_sim_substrate_v2","predict_da_theta",
                                 "fit_stage1","fit_da_pca"), exists)), "\n")
cat("simulate_counts is exp version:", any(grepl("exp\\(", deparse(body(simulate_counts)))), "\n")
print(grep("\\.rds$", tar22, value = TRUE))

Warning message in readChar(con, 5L, useBytes = TRUE):
“cannot open compressed file '/content/saves_pilot/pilot_session.RData', probable reason 'No such file or directory'”


ERROR: Error in readChar(con, 5L, useBytes = TRUE): cannot open the connection


In [ ]:
list.files("/content")

[1] "drive"        "sample_data"  "saves"        "site-library"

In [ ]:
rdata <- list.files("/content/saves", pattern = "\\.RData$",
                    recursive = TRUE, full.names = TRUE)
print(rdata)
pre <- ls()
load(rdata[1])
loaded <- setdiff(ls(), c(pre, "pre", "rdata"))

source(file.path(DRIVE, "fns.R"))

tar22 <- system(sprintf("tar -tzf '%s/saves_eod_2026-07-22.tar.gz'", DRIVE), intern = TRUE)

cat("objects loaded:", length(loaded), "\n")
cat("L dims:", paste(dim(L), collapse = " x "), "\n")
cat("constants ok:", all(sapply(c("L","lag_weights","study_dates"), exists)), "\n")
cat("fns ok:", all(sapply(c("build_city_sim_substrate_v2","predict_da_theta",
                            "fit_stage1","fit_da_pca"), exists)), "\n")
cat("simulate_counts exp:", any(grepl("exp\\(", deparse(body(simulate_counts)))), "\n")
cat("tar 07-22 rds:", sum(grepl("\\.rds$", tar22)), "\n")

[1] "/content/saves/pilot_session.RData"
objects loaded: 47 
L dims: 17 x 3 
constants ok: TRUE 
fns ok: TRUE 
simulate_counts exp: TRUE 
tar 07-22 rds: 27 


[1] "/content/saves/pilot_session.RData"
objects loaded: 47
L dims: 17 x 3
constants ok: TRUE
fns ok: TRUE
simulate_counts exp: TRUE
tar 07-22 rds: 27

---

restore clean. 47 objs from pilot_session, L 17x3, exp simulate_counts, fns sourced last.  

untar 07-22, list the 27 names

In [ ]:
system(sprintf("tar -xzf '%s/saves_eod_2026-07-22.tar.gz' -C /content", DRIVE))
rds <- list.files("/content/saves_eod", pattern = "\\.rds$", full.names = FALSE)
cat(length(rds), "files\n")
print(sort(rds))

27 files
 [1] "cma_age_data_mtlvan.rds" "cma_predictors.rds"     
 [3] "diag3_v2.rds"            "diag5_v2.rds"           
 [5] "fn_fit_stage1.rds"       "fn_qaic.rds"            
 [7] "fn_reduce_fit.rds"       "ids3_v2.rds"            
 [9] "ids5_v2.rds"             "mtl_substrate.rds"      
[11] "new_age_ottqc.rds"       "red_mtl_fit.rds"        
[13] "red_tor.rds"             "red_van_fit.rds"        
[15] "red3_v2.rds"             "red5_v2.rds"            
[17] "reduced_df.rds"          "reduced_mtl.rds"        
[19] "reduced_tor.rds"         "reduced_van.rds"        
[21] "res5_v2.rds"             "stage2_k5_v2.rds"       
[23] "stage2.rds"              "van_substrate.rds"      
[25] "vcov_list.rds"           "Z3_v2.rds"              
[27] "Z5_v2.rds"              


27 files
 [1] "cma_age_data_mtlvan.rds" "cma_predictors.rds"     
 [3] "diag3_v2.rds"            "diag5_v2.rds"           
 [5] "fn_fit_stage1.rds"       "fn_qaic.rds"            
 [7] "fn_reduce_fit.rds"       "ids3_v2.rds"            
 [9] "ids5_v2.rds"             "mtl_substrate.rds"      
[11] "new_age_ottqc.rds"       "red_mtl_fit.rds"        
[13] "red_tor.rds"             "red_van_fit.rds"        
[15] "red3_v2.rds"             "red5_v2.rds"            
[17] "reduced_df.rds"          "reduced_mtl.rds"        
[19] "reduced_tor.rds"         "reduced_van.rds"        
[21] "res5_v2.rds"             "stage2_k5_v2.rds"       
[23] "stage2.rds"              "van_substrate.rds"      
[25] "vcov_list.rds"           "Z3_v2.rds"              
[27] "Z5_v2.rds"              

---



load the P1 candidates, check shapes. Z5 should be ~5-city DA count x 17, ids match rows, stage2 coefs = 20

In [ ]:
Z5   <- readRDS("/content/saves_eod/Z5_v2.rds")
ids5 <- readRDS("/content/saves_eod/ids5_v2.rds")
cmap <- readRDS("/content/saves_eod/cma_predictors.rds")
s2   <- readRDS("/content/saves_eod/stage2_k5_v2.rds")

cat("Z5:", paste(dim(Z5), collapse=" x "), "| NAs:", sum(is.na(Z5)), "\n")
str(ids5, max.level = 1)
cat("ids rows match Z5:", NROW(ids5) == nrow(Z5), "\n")
print(cmap)
cat("s2 class:", class(s2)[1], "\n")

Z5:  | NAs: 0 
List of 5
 $ Toronto  : chr [1:7682] "35180018" "35180019" "35180021" "35180022" ...
 $ Montreal : chr [1:6504] "24520101" "24520102" "24520103" "24520104" ...
 $ Vancouver: chr [1:3573] "59150004" "59150005" "59150006" "59150007" ...
 $ Ottawa   : chr [1:2042] "24800069" "24800070" "24800071" "24800072" ...
 $ Quebec   : chr [1:1310] "24190051" "24190052" "24190053" "24190054" ...
ids rows match Z5:  
         CMA       PC1         PC2        PC3
      <char>     <num>       <num>      <num>
1:   Toronto -1.998866 -0.71173777 -2.9258033
2:  Montreal -3.730649  0.95436513  1.8301977
3: Vancouver  1.354757 -3.38203234  1.6156235
4:    Ottawa  2.233846 -0.01748303 -1.0322194
5:    Quebec  2.140912  3.15688801  0.5122016
s2 class: list 


Z5:  | NAs: 0
List of 5
 $ Toronto  : chr [1:7682] "35180018" "35180019" "35180021" "35180022" ...
 $ Montreal : chr [1:6504] "24520101" "24520102" "24520103" "24520104" ...
 $ Vancouver: chr [1:3573] "59150004" "59150005" "59150006" "59150007" ...
 $ Ottawa   : chr [1:2042] "24800069" "24800070" "24800071" "24800072" ...
 $ Quebec   : chr [1:1310] "24190051" "24190052" "24190053" "24190054" ...
ids rows match Z5:  
         CMA       PC1         PC2        PC3
      <char>     <num>       <num>      <num>
1:   Toronto -1.998866 -0.71173777 -2.9258033
2:  Montreal -3.730649  0.95436513  1.8301977
3: Vancouver  1.354757 -3.38203234  1.6156235
4:    Ottawa  2.233846 -0.01748303 -1.0322194
5:    Quebec  2.140912  3.15688801  0.5122016
s2 class: list  

---

27 rds, Z5/ids5 are per-city lists, not stacked; 5 cities, ids sum 21,111. cma_predictors = 5x3 top-down scores, banked. stage2_k5_v2 is a list. nothing needs regenerating

confirm Z5 = list of 5 city matrices, each n x 17 matching ids. stack in fixed order tor/mtl/van/ott/que → Zs 21111 x 17, ids_all, city_all

In [ ]:
cities <- c("Toronto","Montreal","Vancouver","Ottawa","Quebec")
stopifnot(identical(names(Z5), cities), identical(names(ids5), cities))
for (cc in cities) stopifnot(nrow(Z5[[cc]]) == length(ids5[[cc]]),
                             ncol(Z5[[cc]]) == 17)
Zs       <- do.call(rbind, Z5[cities])
ids_all  <- unlist(ids5[cities], use.names = FALSE)
city_all <- rep(cities, times = sapply(Z5[cities], nrow))

cat("Zs:", paste(dim(Zs), collapse=" x "), "| NAs:", sum(is.na(Zs)), "\n")
cat("ids:", length(ids_all), "| dup ids:", sum(duplicated(ids_all)), "\n")
print(table(city_all))

Zs: 21111 x 17 | NAs: 0 
ids: 21111 | dup ids: 0 
city_all
 Montreal    Ottawa    Quebec   Toronto Vancouver 
     6504      2042      1310      7682      3573 


Zs: 21111 x 17 | NAs: 0
ids: 21111 | dup ids: 0
city_all
 Montreal    Ottawa    Quebec   Toronto Vancouver
     6504      2042      1310      7682      3573

---

stacked Zs is 21111 x 17, zero NAs, zero duplicate ids; per-city rows: tor 7682, mtl 6504, van 3573, ott 2042, que 1310.

stack is aligned by construction: per-city assertions passed before the rbind, and zero dups means no DA rides in two cities. This matrix is P1's input; city_all carries the grouping the CMA weighting will need.

fit the bottom-up rotation: prcomp on 21111x17 scaled. da scores native, loadings kept for C3. expect cum var 75-90, scores 21111 x 3, city means separated

In [ ]:
pca_da <- fit_da_pca(Zs, ids_all)
scores <- pca_da$scores
scores[, city := city_all]

cat("scores:", paste(dim(scores), collapse=" x "), "| NAs:", sum(is.na(scores)), "\n")
cat("var explained:", round(pca_da$var_explained, 3), "\n")
print(scores[, lapply(.SD, mean), by = city, .SDcols = c("PC1","PC2","PC3")])

PCA: cum var first 3 = 86.9%
scores: 21111 x 5 | NAs: 0 
var explained: 0.321 0.289 0.259 
        city        PC1        PC2        PC3
      <char>      <num>      <num>      <num>
1:   Toronto -0.4831056 -1.1978792 -0.1178088
2:  Montreal -0.8467696  1.8483489  0.3799827
3: Vancouver  0.7127652 -0.2830405 -1.0951858
4:    Ottawa  1.7419703 -1.0922867  0.2429669
5:    Quebec  2.3777048  0.3222899  1.4126414


PCA: cum var first 3 = 86.9%
scores: 21111 x 5 | NAs: 0
var explained: 0.321 0.289 0.259
        city        PC1        PC2        PC3
      <char>      <num>      <num>      <num>
1:   Toronto -0.4831056 -1.1978792 -0.1178088
2:  Montreal -0.8467696  1.8483489  0.3799827
3: Vancouver  0.7127652 -0.2830405 -1.0951858
4:    Ottawa  1.7419703 -1.0922867  0.2429669
5:    Quebec  2.3777048  0.3222899  1.4126414

---

bottom-up rotation fitted: cum var 86.9%, components split 32/29/26 — equal factors by design, the missing 13% is the column noise. scores 21111 x 5 clean, city means separate ~3 units on PC1.

pc1 city order and signs match the top-down predictors exactly (mtl < tor < van < ott < que both ways), so pc1's column sign was never flipped between levels — if p3 still inverts, the bug lived elsewhere in the projection. tighter spread is expected: means of scores, not scores of means.

In [ ]:
mtl_sub <- readRDS("/content/saves_eod/mtl_substrate.rds")
van_sub <- readRDS("/content/saves_eod/van_substrate.rds")
ottqc   <- readRDS("/content/saves_eod/new_age_ottqc.rds")

cat("session da_age:", exists("da_age"), "\n")
str(da_age, max.level = 1)
str(mtl_sub$da_age, max.level = 1)
str(van_sub$da_age, max.level = 1)
str(ottqc, max.level = 2)

session da_age: TRUE 
Classes ‘data.table’ and 'data.frame':	7694 obs. of  7 variables:
 $ ALT_GEO_CODE: chr  "35180018" "35180019" "35180021" "35180022" ...
 $ age_0_64    : num  1375 550 410 340 400 ...
 $ age_65_74   : num  115 80 50 50 65 70 170 40 165 200 ...
 $ age_75_84   : num  75 55 30 10 15 55 130 10 100 145 ...
 $ age_85p     : num  15 20 5 5 10 30 95 5 30 165 ...
 $ total       : num  1580 705 495 405 490 ...
 $ da_idx      : int  1 2 3 4 5 6 7 8 9 10 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)> 
Classes ‘data.table’ and 'data.frame':	6504 obs. of  7 variables:
 $ ALT_GEO_CODE: chr  "24520101" "24520102" "24520103" "24520104" ...
 $ age_0_64    : num  705 505 390 745 470 370 745 655 510 1050 ...
 $ age_65_74   : num  115 50 35 60 50 65 90 115 85 110 ...
 $ age_75_84   : num  45 15 5 15 25 40 45 70 60 55 ...
 $ age_85p     : num  5 5 0 5 0 15 10 25 35 5 ...
 $ total       : num  870 575 430 825 545 490 890 865 690 1220 ...
 $ da_idx      : int  1 2 3 4 5 6 7 8 9 10 .

session da_age: TRUE
Classes ‘data.table’ and 'data.frame':	7694 obs. of  7 variables:
 $ ALT_GEO_CODE: chr  "35180018" "35180019" "35180021" "35180022" ...
 $ age_0_64    : num  1375 550 410 340 400 ...
 $ age_65_74   : num  115 80 50 50 65 70 170 40 165 200 ...
 $ age_75_84   : num  75 55 30 10 15 55 130 10 100 145 ...
 $ age_85p     : num  15 20 5 5 10 30 95 5 30 165 ...
 $ total       : num  1580 705 495 405 490 ...
 $ da_idx      : int  1 2 3 4 5 6 7 8 9 10 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)>
Classes ‘data.table’ and 'data.frame':	6504 obs. of  7 variables:
 $ ALT_GEO_CODE: chr  "24520101" "24520102" "24520103" "24520104" ...
 $ age_0_64    : num  705 505 390 745 470 370 745 655 510 1050 ...
 $ age_65_74   : num  115 50 35 60 50 65 90 115 85 110 ...
 $ age_75_84   : num  45 15 5 15 25 40 45 70 60 55 ...
 $ age_85p     : num  5 5 0 5 0 15 10 25 35 5 ...
 $ total       : num  870 575 430 825 545 490 890 865 690 1220 ...
 $ da_idx      : int  1 2 3 4 5 6 7 8 9 10 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)>
Classes ‘data.table’ and 'data.frame':	3573 obs. of  7 variables:
 $ ALT_GEO_CODE: chr  "59150004" "59150005" "59150006" "59150007" ...
 $ age_0_64    : num  240 435 355 365 430 315 640 395 235 825 ...
 $ age_65_74   : num  80 45 55 100 90 45 120 70 30 180 ...
 $ age_75_84   : num  40 35 20 60 40 15 75 35 20 130 ...
 $ age_85p     : num  10 5 15 10 15 10 20 15 10 35 ...
 $ total       : num  370 520 445 535 575 385 855 515 295 1170 ...
 $ da_idx      : int  1 2 3 4 5 6 7 8 9 10 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)>
List of 2
 $ Ottawa:Classes ‘data.table’ and 'data.frame':	2045 obs. of  6 variables:
  ..$ ALT_GEO_CODE: chr [1:2045] "24800069" "24800070" "24800071" "24800072" ...
  ..$ age_0_64    : num [1:2045] 355 1195 275 440 230 ...
  ..$ age_65_74   : num [1:2045] 50 120 65 60 70 65 90 100 205 90 ...
  ..$ age_75_84   : num [1:2045] 25 45 35 35 30 15 40 40 85 30 ...
  ..$ age_85p     : num [1:2045] 5 15 15 15 15 20 5 5 15 5 ...
  ..$ total       : num [1:2045] 435 1375 390 550 345 ...
  ..- attr(*, ".internal.selfref")=<pointer: (nil)>
 $ Quebec:Classes ‘data.table’ and 'data.frame':	1317 obs. of  6 variables:
  ..$ ALT_GEO_CODE: chr [1:1317] "24190051" "24190052" "24190053" "24190054" ...
  ..$ age_0_64    : num [1:1317] 675 340 755 500 630 ...
  ..$ age_65_74   : num [1:1317] 100 75 80 80 85 110 130 60 60 130 ...
  ..$ age_75_84   : num [1:1317] 65 30 90 30 30 60 50 50 25 60 ...
  ..$ age_85p     : num [1:1317] 15 5 115 10 5 25 10 20 0 10 ...
  ..$ total       : num [1:1317] 855 450 1040 620 750 ...
  ..- attr(*, ".internal.selfref")=<pointer: (nil)>

  --

  population tables live in four places, all with ALT_GEO_CODE + total: session da_age 7694, mtl 6504, van 3573, ottqc 2045/1317.

tor/ott/que run long by 12/3/7 — pre-water-drop vintages. match on DAUID absorbs them; unmatched must land at exactly 0 after the join or the id universes drifted.

stack the five pop tables to DAUID+total, match score ids in, hang weights, weighted city means. expect 0 unmatched, means ≈ block 4

In [ ]:
pop_all <- rbind(
  da_age[,            .(DAUID = ALT_GEO_CODE, total)],
  mtl_sub$da_age[,    .(DAUID = ALT_GEO_CODE, total)],
  van_sub$da_age[,    .(DAUID = ALT_GEO_CODE, total)],
  ottqc$Ottawa[,      .(DAUID = ALT_GEO_CODE, total)],
  ottqc$Quebec[,      .(DAUID = ALT_GEO_CODE, total)]
)
m <- match(ids_all, pop_all$DAUID)
cat("unmatched:", sum(is.na(m)), "\n")
scores[, w := pop_all$total[m]]
cat("w range:", range(scores$w), "| NAs:", sum(is.na(scores$w)), "\n")

cma_bu <- scores[, .(PC1 = weighted.mean(PC1, w),
                     PC2 = weighted.mean(PC2, w),
                     PC3 = weighted.mean(PC3, w)), by = city]
print(cma_bu)

unmatched: 0 
w range: 15 29665 | NAs: 0 
        city        PC1        PC2        PC3
      <char>      <num>      <num>      <num>
1:   Toronto -0.4687235 -1.1850506 -0.1369552
2:  Montreal -0.8649134  1.8452119  0.4024157
3: Vancouver  0.6855112 -0.2786731 -1.1095281
4:    Ottawa  1.6750934 -1.1459279  0.2206457
5:    Quebec  2.3863511  0.3430853  1.4153834


unmatched: 0
w range: 15 29665 | NAs: 0
        city        PC1        PC2        PC3
      <char>      <num>      <num>      <num>
1:   Toronto -0.4687235 -1.1850506 -0.1369552
2:  Montreal -0.8649134  1.8452119  0.4024157
3: Vancouver  0.6855112 -0.2786731 -1.1095281
4:    Ottawa  1.6750934 -1.1459279  0.2206457
5:    Quebec  2.3863511  0.3430853  1.4153834

---
all 21111 ids matched, weights 15 to 29665, no NAs. weighted city means sit within ~0.07 of the unweighted ones, same order throughout.

weighting barely moves the means because the population is independent of the factors by construction, the tilt had nothing to grab. cma_bu is the bottom-up predictor table; P1 done, both halves.

look inside stage2_k5_v2 and red5_v2: find the K=5 theta matrix and V list to refit on  

In [ ]:
str(s2, max.level = 1)
red5 <- readRDS("/content/saves_eod/red5_v2.rds")
str(red5, max.level = 1)

List of 4
 $ fit    :List of 25
  ..- attr(*, "class")= chr "mixmeta"
 $ pred_df:Classes ‘data.table’ and 'data.frame':	5 obs. of  11 variables:
  ..- attr(*, ".internal.selfref")=<pointer: (nil)> 
 $ method : chr "reml"
 $ formula:Class 'formula'  language cbind(theta1, theta2, theta3, theta4, theta5) ~ PC1 + PC2 + PC3
  .. ..- attr(*, ".Environment")=<environment: 0x5948427f9fd0> 
List of 5
 $ Toronto  :List of 6
 $ Montreal :List of 6
 $ Vancouver:List of 6
 $ Ottawa   :List of 6
 $ Quebec   :List of 6


List of 4
 $ fit    :List of 25
  ..- attr(*, "class")= chr "mixmeta"
 $ pred_df:Classes ‘data.table’ and 'data.frame':	5 obs. of  11 variables:
  ..- attr(*, ".internal.selfref")=<pointer: (nil)>
 $ method : chr "reml"
 $ formula:Class 'formula'  language cbind(theta1, theta2, theta3, theta4, theta5) ~ PC1 + PC2 + PC3
  .. ..- attr(*, ".Environment")=<environment: 0x5948427f9fd0>
List of 5
 $ Toronto  :List of 6
 $ Montreal :List of 6
 $ Vancouver:List of 6
 $ Ottawa   :List of 6
 $ Quebec   :List of 6

 ---

s2 = fit + pred_df + method + formula, no theta/V. Inputs live in red5 instead, five crossreduce lists of 6.

formula confirmed cbind(5 thetas) ~ 3 PCs. refit = same pool, new predictor values from cma_bu

pull theta_star/V_star from red5 in city order, align cma_bu, refit mixmeta reml. expect converged, 20 coefs, df.resid −10

In [ ]:
stopifnot(all(c("theta_star","V_star") %in% names(red5$Toronto)))
theta_bu <- t(sapply(red5[cities], function(r) r$theta_star))
V_bu     <- lapply(red5[cities], function(r) r$V_star)
setkey(cma_bu, NULL); cma_bu <- cma_bu[match(cities, city)]
stopifnot(identical(cma_bu$city, cities), nrow(theta_bu) == 5)

stage2_bu <- mixmeta(theta_bu ~ PC1 + PC2 + PC3, S = V_bu,
                     data = cma_bu, method = "reml")

cat("converged:", stage2_bu$converged, "\n")
cat("coefs:", length(coef(stage2_bu)), "| df.resid:", stage2_bu$df.residual, "\n")
print(round(matrix(coef(stage2_bu), nrow = 5), 3))

converged: TRUE 
coefs: 20 | df.resid: -10 
       [,1]   [,2]  [,3]   [,4]
[1,]  0.562 -2.043 1.603  0.625
[2,] -4.297 -0.326 1.725  0.385
[3,] -3.580 -0.959 1.739 -0.616
[4,] -4.325  0.750 0.904  0.486
[5,]  0.827 -1.940 2.172 -2.084


converged: TRUE
coefs: 20 | df.resid: -10
       [,1]   [,2]  [,3]   [,4]
[1,]  0.562 -2.043 1.603  0.625
[2,] -4.297 -0.326 1.725  0.385
[3,] -3.580 -0.959 1.739 -0.616
[4,] -4.325  0.750 0.904  0.486
[5,]  0.827 -1.940 2.172 -2.084

---

refit converged, 20 coefs, df.resid −10 as expected at K=5. coefficient matrix in the bottom-up coordinate system, no banked comparison meaningful

pc columns nonzero and mixed-sign, predictors active in the pool. wWhether the map comes out right-side-up is P3's number, nothing here can tell us yet

toronto native scores through the new coefs, log-RR at 27.19 vs truth composite. expect 7682 distinct rows, pearson positive ~0.9. negative = stop

In [ ]:
stopifnot(exists("truth_factors"))
tor_sc <- as.matrix(scores[city == "Toronto", .(PC1, PC2, PC3)])
tor_id <- scores[city == "Toronto", DAUID]
stopifnot(identical(tor_id, ids5$Toronto))

th_bu <- predict_da_theta(stage2_bu, tor_sc, tor_id)
cat("distinct rows:", nrow(unique(th_bu[, .(theta1,theta2,theta3,theta4,theta5)])), "\n")
print(names(red5$Toronto))
str(truth_factors, max.level = 1)

distinct rows: 7682 
[1] "cma"         "age"         "theta_star"  "V_star"      "cen"        
[6] "reduced_obj"
Classes ‘data.table’ and 'data.frame':	7694 obs. of  5 variables:
 $ DAUID : chr  "35180018" "35180019" "35180021" "35180022" ...
 $ da_idx: int  1 2 3 4 5 6 7 8 9 10 ...
 $ F1    : num  1.371 -0.565 0.363 0.633 0.404 ...
 $ F2    : num  -0.23 1.875 -1.587 -1.278 -0.165 ...
 $ F3    : num  2.502 0.837 0.177 1.7 -1.135 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)> 


distinct rows: 7682
[1] "cma"         "age"         "theta_star"  "V_star"      "cen"        
[6] "reduced_obj"
Classes ‘data.table’ and 'data.frame':	7694 obs. of  5 variables:
 $ DAUID : chr  "35180018" "35180019" "35180021" "35180022" ...
 $ da_idx: int  1 2 3 4 5 6 7 8 9 10 ...
 $ F1    : num  1.371 -0.565 0.363 0.633 0.404 ...
 $ F2    : num  -0.23 1.875 -1.587 -1.278 -0.165 ...
 $ F3    : num  2.502 0.837 0.177 1.7 -1.135 ...
 - attr(*, ".internal.selfref")=<pointer: (nil)>


---

th_bu built from native scores, 7682 distinct rows. red5 city lists carry cma, age, theta_star, V_star, cen, reduced_obj, no flat basis object. truth_factors is the 7694 pre-drop vintage, match by DAUID absorbs it

match truth to tor ids by DAUID, correlate predicted heat contrast with 0.4F1+0.2F3. basis-free: logRR at fixed temp is a fixed linear combo of theta, proportionality cancels in cor. expect positive ~0.9

In [ ]:
tm <- match(tor_id, truth_factors$DAUID)
cat("unmatched:", sum(is.na(tm)), "\n")
truth <- with(truth_factors[tm], 0.4*F1 + 0.2*F3)

th_mat <- as.matrix(th_bu[, .(theta1,theta2,theta3,theta4,theta5)])
print(round(cor(th_mat, truth)[,1], 3))

unmatched: 0 
theta1 theta2 theta3 theta4 theta5 
-0.008 -0.013 -0.014 -0.006 -0.011 


unmatched: 0
theta1 theta2 theta3 theta4 theta5
-0.008 -0.013 -0.014 -0.006 -0.011

---

truth matched, five theta correlations against truth all negative similar size

undetermined by design: theta signs live in the basis convention, one shared scalar drives all five, the map's sign only shows through the curve not the coefficients

argvar off red_tor's cb_template, onebasis, weights = b(27.19) - b(cen), composite logRR vs truth. positive ~0.9

In [ ]:
red_tor <- readRDS("/content/saves_eod/red_tor.rds")
av  <- attr(red_tor$cb_template, "argvar")
cen <- red5$Toronto$cen
cat("cen:", cen, "| knots:", round(av$knots, 2), "\n")

b   <- onebasis(c(27.19, cen), fun = av$fun, degree = av$degree, knots = av$knots)
wts <- b[1, ] - b[2, ]
logrr_hat <- as.vector(th_mat %*% wts)

cat("logrr sd:", round(sd(logrr_hat), 3), "\n")
cat("pearson:", round(cor(logrr_hat, truth), 3),
    "| spearman:", round(cor(logrr_hat, truth, method = "spearman"), 3), "\n")

cen: 19.38333 | knots: 12.51 22.18 24.27 
logrr sd: 5.655 
pearson: -0.007 | spearman: 0.002 


cen: 19.38333 | knots: 12.51 22.18 24.27
logrr sd: 5.655
pearson: -0.007 | spearman: 0.002

---

pearson −0.007, sd 5.655onebasis on two points defaulted boundary knots to [19.4, 27.2] instead of the full temp range, one knot outside boundary, wts in the wrong coordinate system

verdict still pending, fix is passing the boundary explicitly

boundary knots passed explicitly from argvar. expect boundary ~3.7 to 30, pearson positive ~0.9, sd back near 1.3

In [ ]:
print(names(av))
bk <- av$Boundary.knots
cat("boundary:", bk, "\n")

b2  <- onebasis(c(27.19, cen), fun = av$fun, degree = av$degree,
                knots = av$knots, Boundary.knots = bk)
wts2 <- b2[1, ] - b2[2, ]
logrr2 <- as.vector(th_mat %*% wts2)

cat("logrr sd:", round(sd(logrr2), 3), "\n")
cat("pearson:", round(cor(logrr2, truth), 3),
    "| spearman:", round(cor(logrr2, truth, method = "spearman"), 3), "\n")

[1] "fun"            "knots"          "degree"         "intercept"     
[5] "Boundary.knots"
boundary: 3.68 30.085 
logrr sd: 1.226 
pearson: 0.006 | spearman: 0.009 


[1] "fun"            "knots"          "degree"         "intercept"     
[5] "Boundary.knots"
boundary: 3.68 30.085
logrr sd: 1.226
pearson: 0.006 | spearman: 0.009

boundary read off argvar: 3.68 to 30.09, dot on the temp range from banked Tor's temp range, sd drops 5.655 to 1.226, pearson 0.006, spearman 0.009

the ruler is fixed, the amplitude came home, and the correlation is genuinely zero not flipped. Zero is a different animal than −0.9: a flip preserves ordering, zero says the two vectors don't know each other at all

regenerate toronto v2 at seed 77, real truth. expect mu −0.33 0.655 0.384, n_da 7682, Z identical to Z5 toronto

In [ ]:
tor_temp <- read_daymet(file.path(DRIVE, "toronto_daymet_2015_2019.csv"))
tor_sub  <- build_city_sim_substrate_v2(tor_temp, da_age, L, seed = 77)

cat("n_da:", tor_sub$n_da, "\n")
cat("mu:", round(tor_sub$mu, 3), "\n")
cat("Z identical to Z5:", isTRUE(all.equal(tor_sub$Z, Z5$Toronto,
                                           check.attributes = FALSE)), "\n")
cat("ids match:", identical(tor_sub$truth_factors$DAUID, tor_id), "\n")

n_da: 7682 
mu: -0.33 0.655 0.384 
Z identical to Z5: TRUE 
ids match: TRUE 


n_da: 7682
mu: -0.33 0.655 0.384
Z identical to Z5: TRUE
ids match: TRUE

===

seed 77 regeneration: n_da 7682, mu −0.33 0.655 0.384, Z bit-identical to the stacked Z5 toronto, ids already in score order

all four fingerprints hit, so Z5 is confirmed seed-77 vintage and the session truth_factors was the v1 imposter. real truth is tor_sub's no match step needed

verdict on matched vintages: same wts2, truth from tor_sub, expect pearson positive ~0.9

In [ ]:
truth2 <- with(tor_sub$truth_factors, 0.4*F1 + 0.2*F3)

cat("pearson:", round(cor(logrr2, truth2), 3),
    "| spearman:", round(cor(logrr2, truth2, method = "spearman"), 3), "\n")
cat("truth sd:", round(sd(truth2), 3), "| logrr sd:", round(sd(logrr2), 3), "\n")

pearson: -0.946 | spearman: -0.94 
truth sd: 0.445 | logrr sd: 1.226 


pearson: -0.946 | spearman: -0.94
truth sd: 0.445 | logrr sd: 1.226

---

pearson −0.946, spearman −0.94, truth sd 0.445, logrr sd 1.226, amplitude ratio 2.76

the flip survived a build with one rotation and native scores. signature identical to the old map, two different rotations, same flip and same inflation, which points at machinery both runs shared downstream of the rotation

diagnosis: print coef names, predict at the 5 city means, compare to fitted. mismatch convicts the reshape in predict_da_theta

In [ ]:
print(names(coef(stage2_bu)))

cma_hat <- predict_da_theta(stage2_bu,
                            as.matrix(cma_bu[, .(PC1, PC2, PC3)]),
                            cma_bu$city)
print(round(as.matrix(cma_hat[, .(theta1,theta2,theta3,theta4,theta5)]) -
            fitted(stage2_bu), 4))

 [1] "b1.(Intercept)" "b2.(Intercept)" "b3.(Intercept)" "b4.(Intercept)"
 [5] "b5.(Intercept)" "b1.PC1"         "b2.PC1"         "b3.PC1"        
 [9] "b4.PC1"         "b5.PC1"         "b1.PC2"         "b2.PC2"        
[13] "b3.PC2"         "b4.PC2"         "b5.PC2"         "b1.PC3"        
[17] "b2.PC3"         "b3.PC3"         "b4.PC3"         "b5.PC3"        
     theta1 theta2 theta3 theta4 theta5
[1,]      0      0      0      0      0
[2,]      0      0      0      0      0
[3,]      0      0      0      0      0
[4,]      0      0      0      0      0
[5,]      0      0      0      0      0


[1] "b1.(Intercept)" "b2.(Intercept)" "b3.(Intercept)" "b4.(Intercept)"
 [5] "b5.(Intercept)" "b1.PC1"         "b2.PC1"         "b3.PC1"        
 [9] "b4.PC1"         "b5.PC1"         "b1.PC2"         "b2.PC2"        
[13] "b3.PC2"         "b4.PC2"         "b5.PC2"         "b1.PC3"        
[17] "b2.PC3"         "b3.PC3"         "b4.PC3"         "b5.PC3"        
     theta1 theta2 theta3 theta4 theta5
[1,]      0      0      0      0      0
[2,]      0      0      0      0      0
[3,]      0      0      0      0      0
[4,]      0      0      0      0      0
[5,]      0      0      0      0      0

---

coef names b1..b5 per predictor block, reshape matches, prediction at the 5 city means reproduces fitted() to zero

stage 3 machinery is clean, the flip is upstream of prediction: either in how scores relate to truth within a city, or in the stage 2 slope fitted across 5 city means

split the fork: within-toronto cor of each PC with truth, then the effective slope a = wts2 B, then the sign of their alignment

In [ ]:
B <- matrix(coef(stage2_bu), nrow = 5)
a <- as.vector(wts2 %*% B[, 2:4])
cv <- cor(tor_sc, truth2)[, 1]

cat("a:", round(a, 3), "\n")
cat("cor(PC, truth):", round(cv, 3), "\n")
cat("alignment a.cv:", round(sum(a * cv), 3), "\n")
cat("check: cor(const+a.s, logrr2):", round(cor(as.vector(tor_sc %*% a), logrr2), 4), "\n")

a: 0.424 -0.349 -0.343 
cor(PC, truth): -0.557 0.595 0.418 
alignment a.cv: -0.588 
check: cor(const+a.s, logrr2): 1 


a: 0.424 -0.349 -0.343
cor(PC, truth): -0.557 0.595 0.418
alignment a.cv: -0.588
check: cor(const+a.s, logrr2): 1

---

effective slope a (0.424, −0.349, −0.343), within-toronto truth correlations (−0.557, 0.595, 0.418), alignment −0.588 of a possible 0.594, compression check exact

a is antiparallel to the truth direction at 99% of max, a coherent global reversal, not component noise. two candidates left: the realized between-city geometry genuinely points backwards, or the theta rows are paired to the wrong cities, which also fakes any angle

In [ ]:
cat("labels:", sapply(red5[cities], `[[`, "cma"), "\n")

city_hot <- as.vector(theta_bu %*% wts2)
tab <- data.table(cma_bu[, .(city, PC1, PC2, PC3)], hot = round(city_hot, 3))
print(tab)
cat("toronto mean truth:", round(mean(truth2), 3), "\n")

labels: Toronto Montreal Vancouver Ottawa Quebec 
        city        PC1        PC2        PC3   hot
      <char>      <num>      <num>      <num> <num>
1:   Toronto -0.4687235 -1.1850506 -0.1369552 0.398
2:  Montreal -0.8649134  1.8452119  0.4024157 0.290
3: Vancouver  0.6855112 -0.2786731 -1.1095281 1.172
4:    Ottawa  1.6750934 -1.1459279  0.2206457 3.392
5:    Quebec  2.3863511  0.3430853  1.4153834 0.488
toronto mean truth: -0.054 


labels: Toronto Montreal Vancouver Ottawa Quebec
        city        PC1        PC2        PC3   hot
      <char>      <num>      <num>      <num> <num>
1:   Toronto -0.4687235 -1.1850506 -0.1369552 0.398
2:  Montreal -0.8649134  1.8452119  0.4024157 0.290
3: Vancouver  0.6855112 -0.2786731 -1.1095281 1.172
4:    Ottawa  1.6750934 -1.1459279  0.2206457 3.392
5:    Quebec  2.3863511  0.3430853  1.4153834 0.488
toronto mean truth: -0.054

---

labels inside red5 match the list names, pairing clean. city table printed with each city's hot contrast, toronto mean truth −0.054

pairing is dead but the hot column is contaminated

vintage-check mtl/van by Z identity to the stack, then mean truth per city from mu, table against hot

In [ ]:
cat("mtl Z match:", isTRUE(all.equal(mtl_sub$Z, Z5$Montreal, check.attributes = FALSE)),
    "| van Z match:", isTRUE(all.equal(van_sub$Z, Z5$Vancouver, check.attributes = FALSE)), "\n")
mt <- function(s) 0.4*s$mu[1] + 0.2*s$mu[3]
cat("mu mtl:", round(mtl_sub$mu, 3), "| mu van:", round(van_sub$mu, 3), "\n")
print(data.table(city = c("Toronto","Montreal","Vancouver"),
                 mean_truth = round(c(-0.054, mt(mtl_sub), mt(van_sub)), 3),
                 hot = c(0.398, 0.290, 1.172)))

mtl Z match: FALSE | van Z match: FALSE 


ERROR: Error in round(mtl_sub$mu, 3): non-numeric argument to mathematical function


clean city curves: each reduced_obj's own fit at its own p99, against mean truth from ssot mu

In [ ]:
mus <- rbind(Toronto  = c(-0.330, 0.655, 0.384),
             Montreal = c( 1.394, 0.130, 0.251),
             Vancouver= c(-0.196, 0.331,-0.405),
             Ottawa   = c(-0.776,-0.280, 0.080),
             Quebec   = c(-0.077,-1.059, 0.200))
print(names(red5$Toronto$reduced_obj))
own_hot <- sapply(cities, function(cc) {
  r <- red5[[cc]]$reduced_obj
  r$fit[which.min(abs(r$predvar - quantile(r$predvar, 0.99)))]
})
print(data.table(city = cities,
                 mean_truth = round(0.4*mus[cities,1] + 0.2*mus[cities,3], 3),
                 own_hot = round(own_hot, 3)))

 [1] "coefficients" "vcov"         "basis"        "type"         "value"       
 [6] "predvar"      "cen"          "lag"          "bylag"        "fit"         
[11] "se"           "RRfit"        "RRlow"        "RRhigh"       "ci.level"    
[16] "model.class"  "model.link"  
        city mean_truth own_hot
      <char>      <num>   <num>
1:   Toronto     -0.055   1.588
2:  Montreal      0.608   4.165
3: Vancouver     -0.159   3.851
4:    Ottawa     -0.294   6.026
5:    Quebec      0.009  -3.893


[1] "coefficients" "vcov"         "basis"        "type"         "value"       
 [6] "predvar"      "cen"          "lag"          "bylag"        "fit"         
[11] "se"           "RRfit"        "RRlow"        "RRhigh"       "ci.level"    
[16] "model.class"  "model.link"  
        city mean_truth own_hot
      <char>      <num>   <num>
1:   Toronto     -0.055   1.588
2:  Montreal      0.608   4.165
3: Vancouver     -0.159   3.851
4:    Ottawa     -0.294   6.026
5:    Quebec      0.009  -3.893

---

reduced_obj carries the full fitted curve, predvar/fit/RRfit. city fits read at range-p99: tor 1.6, mtl 4.2, van 3.9, ott 6.0, que −3.9

unusable readout, range quantile sits past the last knot in the extrapolation tail. magnitudes impossible under the dgp, que negative at heat, all flap. verdict still open

reread each city's curve at its own third knot, its banked p90, plus knot check. expect mtl highest, tor lowest-ish if stage 1 is faithful

In [ ]:
own_p90 <- sapply(cities, function(cc) {
  r  <- red5[[cc]]$reduced_obj
  k3 <- r$basis$knots[3]
  r$fit[which.min(abs(r$predvar - k3))]
})
print(data.table(city = cities,
                 mean_truth = round(0.4*mus[cities,1] + 0.2*mus[cities,3], 3),
                 knot3 = round(sapply(cities, function(cc) red5[[cc]]$reduced_obj$basis$knots[3]), 2),
                 fit_p90 = round(own_p90, 3)))

ERROR: Error in r$basis$knots: $ operator is invalid for atomic vectors
